legacy/tools/witx-docs.md

```diff
@@ -1,30 +1,90 @@
-sections.
-
-We expect that eventually we will transition to using the full interface
-types specification, with `witx` having minimal additional features. Until then,
-the goals here are to remain aligned with interface types and other relevant
-WebAssembly standards and proposals wherever practical, and to be an input 
-into the design process of interface types.
-
-[module linking]: https://github.com/WebAssembly/module-linking/blob/main/design/proposals/module-linking/Explainer.md
+to existing APIs and proposing new APIs.
+
+The WASI Subgroup is [migrating] away from `witx` and toward [Wit], because
+Wit provides much better support for non-C-like languages, better support for
+API virtualization, it has a path to integrating async behavior into WASI
+APIs in a comprehensive way, and it supports much more expression APIs, such
+as the ability to have `string`s and other types as return types in addition
+to just argument types. At this point, the tooling for Wit is also a lot more
+sophisticated and the [Wit language] and [Canonical ABI] have much more
+documentation.
+
+This document focused on the witx format.
+
+## Return types
+
+Function declarations in witx can have a special `expected` type, which is
+a variant which represents either success or failure, and can return a
+specific type off value for each.
+
+For example, the `fd_read` function declaration in Preview1 contains this:
+
+```witx
+    (result $error (expected $size (error $errno)))
+```
+
+This declares a result named `$error` which returns a value with type
+`$size` on success, and a value with type `$errno` on failure.
+
+The `expected` mechanism assumes that the error value is an enum where 0
+indicates success, and as such it doesn't return an explicit descriminant
+value. In the ABI, the `error` type is returned as the return value and
+the success value is handled by adding an argument of pointer type for
+the function to store the result into.
+
+The resulting ABI for `fd_read` looks like this:
+
+```c
+__wasi_errno_t __wasi_fd_read(
+    __wasi_fd_t fd,
+    const __wasi_iovec_t *iovs,
+    size_t iovs_len,
+    __wasi_size_t *retptr0
+);
+```
+
+## Pointers
+
+Witx supports two pointer types, `pointer` and `const_pointer`, which represent
+pointers into the exported linear memory named "memory". `const_pointer` in a
+function declaration documents that the function does not use the pointer for
+mutating memory. Similar to C, they can point to either a single value or an
+contiguous array of values.
+
+Pointer arguments use the `@witx` syntax inspired by the [annotations proposal].
+
+For example, the `poll_oneoff` function has these arguments:
+
+```witx
+   (param $in (@witx const_pointer $subscription))
+   (param $out (@witx pointer $event))
+```
+
+Pointer values are expected to be aligned, to the alignment of their pointee
+type. If a misaligned pointer is passed to a function, the function shall fail
+with an `errno::inval` error code.
```

Comment:
i'd suggest s/shall/can/ to align the existing implementations and wasm `load/store` semantics.

Reply:
well, actually i prefer to just accept misaligned addresses as wasm `load/store` does.

even if we choose to reject them, i think it should be with a trap (as in eg. [wasm threads](https://github.com/WebAssembly/threads/blob/main/proposals/threads/Overview.md#alignment)), not `inval`.

Reply:
That sounds good to me. I've now changed it to say trapping.

Reply:
will preview2 have the same trapping semantics?

Reply:
Yes; as currently proposed, the [CanonicalABI has](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#loading) traps for both alignment and bounds.